types/vue.d.ts

```diff
@@ -20,30 +24,27 @@ export type CreateElement = {
   (tag: string, data?: VNodeData, children?: VNodeChildren): VNode;
 
   // component constructor or options
-  (tag: Component, children: VNodeChildren): VNode;
-  (tag: Component, data?: VNodeData, children?: VNodeChildren): VNode;
+  (tag: Component<any, any, any, any>, children: VNodeChildren): VNode;
+  (tag: Component<any, any, any, any>, data?: VNodeData, children?: VNodeChildren): VNode;
 
   // async component
-  (tag: AsyncComponent, children: VNodeChildren): VNode;
-  (tag: AsyncComponent, data?: VNodeData, children?: VNodeChildren): VNode;
+  (tag: AsyncComponent<any, any, any, any>, children: VNodeChildren): VNode;
+  (tag: AsyncComponent<any, any, any, any>, data?: VNodeData, children?: VNodeChildren): VNode;
 }
 
-export declare class Vue {
-
-  constructor(options?: ComponentOptions<Vue>);
-
-  $data: Object;
+export interface Vue {
   readonly $el: HTMLElement;
-  readonly $options: ComponentOptions<this>;
+  readonly $options: ComponentOptions<any, any, any, any>;
   readonly $parent: Vue;
   readonly $root: Vue;
   readonly $children: Vue[];
-  readonly $refs: { [key: string]: Vue | Element | Vue[] | Element[]};
+  readonly $refs: { [key: string]: Vue | Element | Vue[] | Element[] };
   readonly $slots: { [key: string]: VNode[] };
   readonly $scopedSlots: { [key: string]: ScopedSlot };
   readonly $isServer: boolean;
+  readonly $data: Record<string, any>;
+  readonly $props: Record<string, any>;
```

Comment:
We can also make Vue generic so `$data` and `$props` can be typed. But I wonder if it is an overkill.